File: src/primitives/hex_test.zig
const std = @import("std");
const testing = std.testing;
const Hex = @import("hex.zig");

// Test hex validation
test "is_valid_hex" {
    // Valid hex strings
    try testing.expect(Hex.is_valid_hex("0x"));
    try testing.expect(Hex.is_valid_hex("0x0"));
    try testing.expect(Hex.is_valid_hex("0x00"));
    try testing.expect(Hex.is_valid_hex("0x0123456789abcdef"));
    try testing.expect(Hex.is_valid_hex("0x0123456789ABCDEF"));
    try testing.expect(Hex.is_valid_hex("0xdeadbeef"));
    
    // Invalid hex strings
    try testing.expect(!Hex.is_valid_hex(""));
    try testing.expect(!Hex.is_valid_hex("0"));
    try testing.expect(!Hex.is_valid_hex("00"));
    try testing.expect(!Hex.is_valid_hex("0xg"));
    try testing.expect(!Hex.is_valid_hex("0x0123456789abcdefg"));
    try testing.expect(!Hex.is_valid_hex("0x "));
    try testing.expect(!Hex.is_valid_hex(" 0x00"));
    try testing.expect(!Hex.is_valid_hex("0x00 "));
}

// Test from bytes
test "from_bytes basic" {
    const allocator = testing.allocator;
    
    // Empty bytes
    const empty = try Hex.from_bytes(allocator, &[_]u8{});
    defer allocator.free(empty);
    try testing.expectEqualStrings("0x", empty);
    
    // Single byte
    const single = try Hex.from_bytes(allocator, &[_]u8{0x61});
    defer allocator.free(single);
    try testing.expectEqualStrings("0x61", single);
    
    // Multiple bytes
    const multiple = try Hex.from_bytes(allocator, &[_]u8{ 0x61, 0x62, 0x63 });
    defer allocator.free(multiple);
    try testing.expectEqualStrings("0x616263", multiple);
    
    // "Hello World!"
    const hello = try Hex.from_bytes(allocator, "Hello World!");
    defer allocator.free(hello);
    try testing.expectEqualStrings("0x48656c6c6f20576f726c6421", hello);
}

test "from_bytes with specific case" {
    const allocator = testing.allocator;
    
    const bytes = [_]u8{ 0xde, 0xad, 0xbe, 0xef };
    
    // Lowercase (default)
    const lower = try Hex.from_bytes(allocator, &bytes);
    defer allocator.free(lower);
    try testing.expectEqualStrings("0xdeadbeef", lower);
    
    // Uppercase
    const upper = try Hex.from_bytes_upper(allocator, &bytes);
    defer allocator.free(upper);
    try testing.expectEqualStrings("0xDEADBEEF", upper);
}

// Test to bytes
test "to_bytes basic" {
    const allocator = testing.allocator;
    
    // Empty hex
    const empty = try Hex.to_bytes(allocator, "0x");
    defer allocator.free(empty);
    try testing.expectEqual(@as(usize, 0), empty.len);
    
    // Single byte
    const single = try Hex.to_bytes(allocator, "0x61");
    defer allocator.free(single);
    try testing.expectEqualSlices(u8, &[_]u8{0x61}, single);
    
    // Multiple bytes
    const multiple = try Hex.to_bytes(allocator, "0x616263");
    defer allocator.free(multiple);
    try testing.expectEqualSlices(u8, &[_]u8{ 0x61, 0x62, 0x63 }, multiple);
    
    // Mixed case
    const mixed = try Hex.to_bytes(allocator, "0xDeAdBeEf");
    defer allocator.free(mixed);
    try testing.expectEqualSlices(u8, &[_]u8{ 0xde, 0xad, 0xbe, 0xef }, mixed);
}

test "to_bytes odd length" {
    const allocator = testing.allocator;
    
    // Odd length should pad with leading zero
    const odd = try Hex.to_bytes(allocator, "0x1");
    defer allocator.free(odd);
    try testing.expectEqualSlices(u8, &[_]u8{0x01}, odd);
    
    const odd2 = try Hex.to_bytes(allocator, "0x123");
    defer allocator.free(odd2);
    try testing.expectEqualSlices(u8, &[_]u8{ 0x01, 0x23 }, odd2);
}

test "to_bytes invalid hex" {
    const allocator = testing.allocator;
    
    // Missing 0x prefix
    const result1 = Hex.to_bytes(allocator, "deadbeef");
    try testing.expectError(Hex.HexError.MissingPrefix, result1);
    
    // Invalid character
    const result2 = Hex.to_bytes(allocator, "0xdeadbeeg");
    try testing.expectError(error.InvalidCharacter, result2);
}

// Test from/to conversions with various types
test "from_u256" {
    const allocator = testing.allocator;
    
    // Zero
    const zero = try Hex.from_u256(allocator, 0);
    defer allocator.free(zero);
    try testing.expectEqualStrings("0x0", zero);
    
    // Small number
    const small = try Hex.from_u256(allocator, 69420);
    defer allocator.free(small);
    try testing.expectEqualStrings("0x10f2c", small);
    
    // Large number
    const large = try Hex.from_u256(allocator, 0xdeadbeef);
    defer allocator.free(large);
    try testing.expectEqualStrings("0xdeadbeef", large);
    
    // Max u256
    const max = try Hex.from_u256(allocator, std.math.maxInt(u256));
    defer allocator.free(max);
    try testing.expectEqualStrings("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff", max);
}

test "to_u256" {
    // Zero
    const zero = try Hex.to_u256("0x0");
    try testing.expectEqual(@as(u256, 0), zero);
    
    // Small number
    const small = try Hex.to_u256("0x10f2c");
    try testing.expectEqual(@as(u256, 69420), small);
    
    // Large number
    const large = try Hex.to_u256("0xdeadbeef");
    try testing.expectEqual(@as(u256, 0xdeadbeef), large);
    
    // Max u256
    const max = try Hex.to_u256("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
    try testing.expectEqual(std.math.maxInt(u256), max);
}

test "to_u256 overflow" {
    // This would overflow u256
    const result = Hex.to_u256("0x10000000000000000000000000000000000000000000000000000000000000000");
    try testing.expectError(Hex.HexError.Overflow, result);
}

// Test slice operations
test "slice hex string" {
    const allocator = testing.allocator;
    
    // Slice from beginning
    const slice1 = try Hex.slice(allocator, "0xdeadbeef", 0, 4);
    defer allocator.free(slice1);
    try testing.expectEqualStrings("0xdead", slice1);
    
    // Slice from middle
    const slice2 = try Hex.slice(allocator, "0xdeadbeefcafe", 4, 8);
    defer allocator.free(slice2);
    try testing.expectEqualStrings("0xbeef", slice2);
    
    // Slice to end
    const slice3 = try Hex.slice(allocator, "0xdeadbeef", 4, null);
    defer allocator.free(slice3);
    try testing.expectEqualStrings("0xbeef", slice3);
}

test "slice bounds checking" {
    const allocator = testing.allocator;
    
    // Start beyond length
    const result1 = Hex.slice(allocator, "0xdead", 10, null);
    try testing.expectError(Hex.HexError.SliceOutOfBounds, result1);
    
    // End before start
    const result2 = Hex.slice(allocator, "0xdead", 4, 2);
    try testing.expectError(Hex.HexError.SliceOutOfBounds, result2);
}

// Test concatenation
test "concat hex strings" {
    const allocator = testing.allocator;
    
    const hex_strings = [_][]const u8{ "0x00", "0x01", "0x02" };
    const concatenated = try Hex.concat(allocator, &hex_strings);
    defer allocator.free(concatenated);
    try testing.expectEqualStrings("0x000102", concatenated);
    
    // With mixed lengths
    const mixed = [_][]const u8{ "0x0", "0x1", "0x69", "0x420" };
    const mixed_concat = try Hex.concat(allocator, &mixed);
    defer allocator.free(mixed_concat);
    try testing.expectEqualStrings("0x0169420", mixed_concat);
}

test "concat with padding" {
    const allocator = testing.allocator;
    
    // Each item padded to 32 bytes
    const hex_strings = [_][]const u8{ "0x01", "0x02", "0x03" };
    const padded = try Hex.concat_padded(allocator, &hex_strings, 32);
    defer allocator.free(padded);
    
    const expected = "0x" ++
        "0000000000000000000000000000000000000000000000000000000000000001" ++
        "0000000000000000000000000000000000000000000000000000000000000002" ++
        "0000000000000000000000000000000000000000000000000000000000000003";
    try testing.expectEqualStrings(expected, padded);
}

// Test padding
test "pad hex left" {
    const allocator = testing.allocator;
    
    const padded = try Hex.pad_left(allocator, "0x123", 4);
    defer allocator.free(padded);
    try testing.expectEqualStrings("0x0123", padded);
    
    const padded2 = try Hex.pad_left(allocator, "0x1", 8);
    defer allocator.free(padded2);
    try testing.expectEqualStrings("0x00000001", padded2);
}

test "pad hex right" {
    const allocator = testing.allocator;
    
    const padded = try Hex.pad_right(allocator, "0x123", 4);
    defer allocator.free(padded);
    try testing.expectEqualStrings("0x1230", padded);
    
    const padded2 = try Hex.pad_right(allocator, "0x1", 8);
    defer allocator.free(padded2);
    try testing.expectEqualStrings("0x10000000", padded2);
}

// Test strip
test "strip leading zeros" {
    const allocator = testing.allocator;
    
    const stripped = try Hex.strip_zeros(allocator, "0x0000123");
    defer allocator.free(stripped);
    try testing.expectEqualStrings("0x123", stripped);
    
    const stripped2 = try Hex.strip_zeros(allocator, "0x00000000");
    defer allocator.free(stripped2);
    try testing.expectEqualStrings("0x0", stripped2);
    
    const no_strip = try Hex.strip_zeros(allocator, "0x123000");
    defer allocator.free(no_strip);
    try testing.expectEqualStrings("0x123000", no_strip);
}

// Test size calculation
test "hex size" {
    try testing.expectEqual(@as(usize, 0), Hex.size("0x"));
    try testing.expectEqual(@as(usize, 1), Hex.size("0x00"));
    try testing.expectEqual(@as(usize, 1), Hex.size("0x01"));
    try testing.expectEqual(@as(usize, 2), Hex.size("0x0001"));
    try testing.expectEqual(@as(usize, 4), Hex.size("0xdeadbeef"));
    try testing.expectEqual(@as(usize, 32), Hex.size("0x0000000000000000000000000000000000000000000000000000000000000000"));
}

// Test comparison
test "hex equality" {
    try testing.expect(Hex.equal("0x01", "0x1"));
    try testing.expect(Hex.equal("0x00", "0x0"));
    try testing.expect(Hex.equal("0xdeadbeef", "0xdeadbeef"));
    try testing.expect(Hex.equal("0xDeAdBeEf", "0xdeadbeef")); // Case insensitive
    
    try testing.expect(!Hex.equal("0x01", "0x02"));
    try testing.expect(!Hex.equal("0x1", "0x10"));
}

// Test common patterns
test "is zero" {
    try testing.expect(Hex.is_zero("0x"));
    try testing.expect(Hex.is_zero("0x0"));
    try testing.expect(Hex.is_zero("0x00"));
    try testing.expect(Hex.is_zero("0x0000"));
    
    try testing.expect(!Hex.is_zero("0x1"));
    try testing.expect(!Hex.is_zero("0x01"));
    try testing.expect(!Hex.is_zero("0x00001"));
}

// Test edge cases
test "empty string handling" {
    const allocator = testing.allocator;
    
    const result = Hex.to_bytes(allocator, "");
    try testing.expectError(Hex.HexError.MissingPrefix, result);
}

test "only prefix" {
    const allocator = testing.allocator;
    
    const bytes = try Hex.to_bytes(allocator, "0x");
    defer allocator.free(bytes);
    try testing.expectEqual(@as(usize, 0), bytes.len);
}

File: src/primitives/hex.test.ts
import { describe, it, expect } from 'vitest';
import { Hex, InvalidHexFormatError, InvalidHexCharacterError, OddLengthHexError } from './hex';

describe('Hex', () => {
  describe('isHex', () => {
    it('validates correct hex strings', () => {
      expect(Hex.isHex('0x0')).toBe(true);
      expect(Hex.isHex('0x00')).toBe(true);
      expect(Hex.isHex('0x1234')).toBe(true);
      expect(Hex.isHex('0xabcdef')).toBe(true);
      expect(Hex.isHex('0xABCDEF')).toBe(true);
      expect(Hex.isHex('0xdeadbeef')).toBe(true);
    });

    it('rejects invalid hex strings', () => {
      expect(Hex.isHex('')).toBe(false);
      expect(Hex.isHex('0x')).toBe(false);
      expect(Hex.isHex('1234')).toBe(false);
      expect(Hex.isHex('0xg')).toBe(false);
      expect(Hex.isHex('0x ')).toBe(false);
      expect(Hex.isHex(' 0x00')).toBe(false);
    });
  });

  describe('validate', () => {
    it('creates validated hex from valid string', () => {
      const hex = Hex.validate.call('0x1234');
      expect(hex).toBe('0x1234');
    });

    it('throws on invalid hex', () => {
      expect(() => Hex.validate.call('1234')).toThrow(InvalidHexFormatError);
      expect(() => Hex.validate.call('0xg')).toThrow(InvalidHexCharacterError);
    });
  });

  describe('fromBytes', () => {
    it('converts empty bytes', () => {
      expect(Hex.fromBytes(new Uint8Array([]))).toBe('0x');
    });

    it('converts single byte', () => {
      expect(Hex.fromBytes(new Uint8Array([0x61]))).toBe('0x61');
    });

    it('converts multiple bytes', () => {
      expect(Hex.fromBytes(new Uint8Array([0x12, 0x34, 0xab, 0xcd]))).toBe('0x1234abcd');
    });

    it('converts text to hex', () => {
      const encoder = new TextEncoder();
      expect(Hex.fromBytes(encoder.encode('Hello World!'))).toBe('0x48656c6c6f20576f726c6421');
    });
  });

  describe('toBytes', () => {
    it('converts empty hex', () => {
      const bytes = Hex.toBytes.call('0x' as Hex);
      expect(bytes.length).toBe(0);
    });

    it('converts single byte', () => {
      const bytes = Hex.toBytes.call('0x61' as Hex);
      expect(Array.from(bytes)).toEqual([0x61]);
    });

    it('converts multiple bytes', () => {
      const bytes = Hex.toBytes.call('0x616263' as Hex);
      expect(Array.from(bytes)).toEqual([0x61, 0x62, 0x63]);
    });

    it('handles mixed case', () => {
      const bytes = Hex.toBytes.call('0xDeAdBeEf' as Hex);
      expect(Array.from(bytes)).toEqual([0xde, 0xad, 0xbe, 0xef]);
    });

    it('throws on odd length', () => {
      expect(() => Hex.toBytes.call('0x1' as Hex)).toThrow(OddLengthHexError);
      expect(() => Hex.toBytes.call('0x123' as Hex)).toThrow(OddLengthHexError);
    });

    it('throws on invalid character', () => {
      expect(() => Hex.toBytes.call('0xdeadbeeg' as Hex)).toThrow(InvalidHexCharacterError);
    });
  });

  describe('concat', () => {
    it('concatenates hex strings', () => {
      const hex1 = '0x1234' as Hex;
      const hex2 = '0xabcd' as Hex;
      expect(Hex.concat(hex1, hex2)).toBe('0x1234abcd');
    });

    it('concatenates multiple hex strings', () => {
      const hex1 = '0x12' as Hex;
      const hex2 = '0x34' as Hex;
      const hex3 = '0xab' as Hex;
      expect(Hex.concat(hex1, hex2, hex3)).toBe('0x1234ab');
    });

    it('concatenates empty hex strings', () => {
      const hex1 = '0x' as Hex;
      const hex2 = '0x1234' as Hex;
      expect(Hex.concat(hex1, hex2)).toBe('0x1234');
    });
  });

  describe('slice', () => {
    it('slices hex string', () => {
      const hex = '0x1234abcd' as Hex;
      expect(Hex.slice.call(hex, 0, 2)).toBe('0x1234');
      expect(Hex.slice.call(hex, 1, 3)).toBe('0x34ab');
    });

    it('slices from start to end', () => {
      const hex = '0x1234abcd' as Hex;
      expect(Hex.slice.call(hex, 2)).toBe('0xabcd');
    });

    it('handles empty slice', () => {
      const hex = '0x1234' as Hex;
      expect(Hex.slice.call(hex, 2, 2)).toBe('0x');
    });
  });

  describe('size', () => {
    it('returns byte size', () => {
      expect(Hex.size.call('0x' as Hex)).toBe(0);
      expect(Hex.size.call('0x12' as Hex)).toBe(1);
      expect(Hex.size.call('0x1234' as Hex)).toBe(2);
      expect(Hex.size.call('0x1234abcd' as Hex)).toBe(4);
    });
  });

  describe('pad', () => {
    it('pads hex to target size', () => {
      const hex = '0x1234' as Hex;
      expect(Hex.pad.call(hex, 4)).toBe('0x00001234');
    });

    it('does not pad if already at target size', () => {
      const hex = '0x1234' as Hex;
      expect(Hex.pad.call(hex, 2)).toBe('0x1234');
    });

    it('does not pad if larger than target size', () => {
      const hex = '0x1234abcd' as Hex;
      expect(Hex.pad.call(hex, 2)).toBe('0x1234abcd');
    });
  });

  describe('trim', () => {
    it('removes leading zeros', () => {
      expect(Hex.trim.call('0x00001234' as Hex)).toBe('0x1234');
    });

    it('removes all leading zeros', () => {
      expect(Hex.trim.call('0x0000' as Hex)).toBe('0x');
    });

    it('does not trim non-zero values', () => {
      expect(Hex.trim.call('0x1234' as Hex)).toBe('0x1234');
    });
  });

  describe('fromNumber / toNumber', () => {
    it('converts zero', () => {
      expect(Hex.fromNumber(0)).toBe('0x0');
      const hex = '0x00' as Hex;
      expect(Hex.toNumber.call(hex)).toBe(0);
    });

    it('converts small numbers', () => {
      expect(Hex.fromNumber(1)).toBe('0x1');
      expect(Hex.fromNumber(15)).toBe('0xf');
      expect(Hex.fromNumber(255)).toBe('0xff');
    });

    it('converts larger numbers', () => {
      expect(Hex.fromNumber(0x1234)).toBe('0x1234');
      expect(Hex.fromNumber(0xdeadbeef)).toBe('0xdeadbeef');
    });

    it('converts with size padding', () => {
      expect(Hex.fromNumber(255, 1)).toBe('0xff');
      expect(Hex.fromNumber(255, 2)).toBe('0x00ff');
      expect(Hex.fromNumber(255, 4)).toBe('0x000000ff');
    });

    it('round-trip number conversions', () => {
      const values = [0, 1, 255, 0x1234, 0xabcdef];
      values.forEach(val => {
        const hex = Hex.fromNumber(val);
        expect(Hex.toNumber.call(hex)).toBe(val);
      });
    });

    it('throws on unsafe integer', () => {
      const hex = '0xffffffffffffffff' as Hex;
      expect(() => Hex.toNumber.call(hex)).toThrow(RangeError);
    });
  });

  describe('fromBigInt / toBigInt', () => {
    it('converts zero', () => {
      expect(Hex.fromBigInt(0n)).toBe('0x0');
      const hex = '0x00' as Hex;
      expect(Hex.toBigInt.call(hex)).toBe(0n);
    });

    it('converts small bigints', () => {
      expect(Hex.fromBigInt(1n)).toBe('0x1');
      expect(Hex.fromBigInt(255n)).toBe('0xff');
      expect(Hex.fromBigInt(0x1234n)).toBe('0x1234');
    });

    it('converts large bigints', () => {
      const large = 0xffffffffffffffffffffffffffffffffn;
      expect(Hex.fromBigInt(large)).toBe('0xffffffffffffffffffffffffffffffff');
    });

    it('converts with size padding', () => {
      expect(Hex.fromBigInt(255n, 1)).toBe('0xff');
      expect(Hex.fromBigInt(255n, 2)).toBe('0x00ff');
      expect(Hex.fromBigInt(255n, 32)).toBe('0x00000000000000000000000000000000000000000000000000000000000000ff');
    });

    it('round-trip bigint conversions', () => {
      const values = [0n, 1n, 255n, 0x1234n, 0xffffffffffffffffn];
      values.forEach(val => {
        const hex = Hex.fromBigInt(val);
        expect(Hex.toBigInt.call(hex)).toBe(val);
      });
    });
  });

  describe('fromBoolean / toBoolean', () => {
    it('converts true to 0x01', () => {
      expect(Hex.fromBoolean(true)).toBe('0x01');
    });

    it('converts false to 0x00', () => {
      expect(Hex.fromBoolean(false)).toBe('0x00');
    });

    it('converts 0x01 to true', () => {
      const hex = '0x01' as Hex;
      expect(Hex.toBoolean.call(hex)).toBe(true);
    });

    it('converts 0x00 to false', () => {
      const hex = '0x00' as Hex;
      expect(Hex.toBoolean.call(hex)).toBe(false);
    });

    it('converts non-zero values to true', () => {
      expect(Hex.toBoolean.call('0xff' as Hex)).toBe(true);
      expect(Hex.toBoolean.call('0x1234' as Hex)).toBe(true);
      expect(Hex.toBoolean.call('0x000001' as Hex)).toBe(true);
    });

    it('converts all-zero values to false', () => {
      expect(Hex.toBoolean.call('0x0000' as Hex)).toBe(false);
      expect(Hex.toBoolean.call('0x00000000' as Hex)).toBe(false);
    });

    it('round-trip boolean conversions', () => {
      expect(Hex.toBoolean.call(Hex.fromBoolean(true))).toBe(true);
      expect(Hex.toBoolean.call(Hex.fromBoolean(false))).toBe(false);
    });
  });

  describe('padRight', () => {
    it('pads hex to target size on right', () => {
      const hex = '0x1234' as Hex;
      expect(Hex.padRight.call(hex, 4)).toBe('0x12340000');
    });

    it('does not pad if already at target size', () => {
      const hex = '0x1234' as Hex;
      expect(Hex.padRight.call(hex, 2)).toBe('0x1234');
    });

    it('does not pad if larger than target size', () => {
      const hex = '0x1234abcd' as Hex;
      expect(Hex.padRight.call(hex, 2)).toBe('0x1234abcd');
    });

    it('pads empty hex', () => {
      const hex = '0x' as Hex;
      expect(Hex.padRight.call(hex, 2)).toBe('0x0000');
    });

    it('pads single byte', () => {
      const hex = '0xff' as Hex;
      expect(Hex.padRight.call(hex, 4)).toBe('0xff000000');
    });
  });

  describe('equals', () => {
    it('compares equal hex strings', () => {
      const hex1 = '0x1234' as Hex;
      expect(Hex.equals.call(hex1, '0x1234' as Hex)).toBe(true);
    });

    it('compares case-insensitively', () => {
      const hex1 = '0xabcd' as Hex;
      expect(Hex.equals.call(hex1, '0xABCD' as Hex)).toBe(true);
      expect(Hex.equals.call(hex1, '0xAbCd' as Hex)).toBe(true);
    });

    it('returns false for different values', () => {
      const hex1 = '0x1234' as Hex;
      expect(Hex.equals.call(hex1, '0x5678' as Hex)).toBe(false);
    });

    it('returns false for different lengths', () => {
      const hex1 = '0x12' as Hex;
      expect(Hex.equals.call(hex1, '0x1234' as Hex)).toBe(false);
    });

    it('compares empty hex strings', () => {
      const hex1 = '0x' as Hex;
      expect(Hex.equals.call(hex1, '0x' as Hex)).toBe(true);
    });
  });

  describe('xor', () => {
    it('performs XOR on same-length hex strings', () => {
      const hex1 = '0x12' as Hex;
      const hex2 = '0x34' as Hex;
      expect(Hex.xor.call(hex1, hex2)).toBe('0x26');
    });

    it('performs XOR with all zeros', () => {
      const hex1 = '0xff' as Hex;
      const hex2 = '0x00' as Hex;
      expect(Hex.xor.call(hex1, hex2)).toBe('0xff');
    });

    it('performs XOR with all ones', () => {
      const hex1 = '0x12' as Hex;
      const hex2 = '0xff' as Hex;
      expect(Hex.xor.call(hex1, hex2)).toBe('0xed');
    });

    it('performs XOR on multi-byte hex', () => {
      const hex1 = '0x1234' as Hex;
      const hex2 = '0xabcd' as Hex;
      expect(Hex.xor.call(hex1, hex2)).toBe('0xb9f9');
    });

    it('XOR with itself returns zeros', () => {
      const hex = '0xdeadbeef' as Hex;
      expect(Hex.xor.call(hex, hex)).toBe('0x00000000');
    });

    it('throws on mismatched lengths', () => {
      const hex1 = '0x12' as Hex;
      const hex2 = '0x1234' as Hex;
      expect(() => Hex.xor.call(hex1, hex2)).toThrow(Hex.InvalidLengthError);
    });
  });

  describe('random', () => {
    it('generates random hex of specified size', () => {
      const hex = Hex.random(4);
      expect(hex.startsWith('0x')).toBe(true);
      expect(Hex.size.call(hex)).toBe(4);
    });

    it('generates different values', () => {
      const hex1 = Hex.random(32);
      const hex2 = Hex.random(32);
      expect(hex1).not.toBe(hex2);
    });

    it('generates valid hex', () => {
      const hex = Hex.random(16);
      expect(Hex.isHex(hex)).toBe(true);
    });

    it('generates zero-size hex', () => {
      const hex = Hex.random(0);
      expect(hex).toBe('0x');
    });

    it('generates single byte', () => {
      const hex = Hex.random(1);
      expect(Hex.size.call(hex)).toBe(1);
    });
  });

  describe('zero', () => {
    it('creates zero-filled hex of specified size', () => {
      expect(Hex.zero(1)).toBe('0x00');
      expect(Hex.zero(2)).toBe('0x0000');
      expect(Hex.zero(4)).toBe('0x00000000');
    });

    it('creates empty hex for size 0', () => {
      expect(Hex.zero(0)).toBe('0x');
    });

    it('creates large zero-filled hex', () => {
      const hex = Hex.zero(32);
      expect(Hex.size.call(hex)).toBe(32);
      expect(hex).toBe('0x0000000000000000000000000000000000000000000000000000000000000000');
    });

    it('creates zero hex that converts to 0', () => {
      const hex = Hex.zero(4);
      expect(Hex.toNumber.call(hex)).toBe(0);
      expect(Hex.toBigInt.call(hex)).toBe(0n);
      expect(Hex.toBoolean.call(hex)).toBe(false);
    });
  });

  describe('round-trip conversions', () => {
    it('maintains data integrity', () => {
      const original = new Uint8Array([0x12, 0x34, 0xab, 0xcd]);
      const hex = Hex.fromBytes(original);
      const result = Hex.toBytes.call(hex);
      expect(Array.from(result)).toEqual(Array.from(original));
    });
  });
});
